File: Vachanam/Audio/AmbientSoundscapePlayer.swift
//
//  AmbientSoundscapePlayer.swift
//  Vachanam
//
//  Ambient focus audio player:
//  - Brown Noise, Pink Noise, 40Hz Binaural Beats, Soft Rain, Library Ambience
//  - Seamless infinite looping
//  - Independent volume control with persistence
//  - Smooth fade-in and fade-out
//  - Optional Study Mode (independent playback when reading is paused)
//

import Foundation
import AVFoundation
import Combine

public enum SoundscapePreset: String, CaseIterable, Identifiable, Codable {
    case brownNoise = "brown_noise"
    case pinkNoise = "pink_noise"
    case binauralFocus = "binaural_focus_40hz"
    case softRain = "soft_rain"
    case libraryAmbience = "library_ambience"
    
    public var id: String { rawValue }
    
    public var title: String {
        switch self {
        case .brownNoise: return "Brown Noise"
        case .pinkNoise: return "Pink Noise"
        case .binauralFocus: return "Binaural Focus"
        case .softRain: return "Soft Rain"
        case .libraryAmbience: return "Library & Café"
        }
    }
    
    public var subtitle: String {
        switch self {
        case .brownNoise: return "Deep, soothing low-frequency rumble"
        case .pinkNoise: return "Balanced waterfall-like calming noise"
        case .binauralFocus: return "40 Hz gamma rhythm for deep focus"
        case .softRain: return "Gentle rainfall with droplet ambience"
        case .libraryAmbience: return "Warm, quiet acoustic atmosphere"
        }
    }
    
    public var systemImage: String {
        switch self {
        case .brownNoise: return "waveform.path"
        case .pinkNoise: return "waveform"
        case .binauralFocus: return "brain.head.profile"
        case .softRain: return "cloud.rain.fill"
        case .libraryAmbience: return "books.vertical.fill"
        }
    }
    
    public var resourceName: String {
        rawValue
    }
}

public class AmbientSoundscapePlayer: NSObject, ObservableObject, AVAudioPlayerDelegate {
    public static let shared = AmbientSoundscapePlayer()
    
    @Published public var currentPreset: SoundscapePreset?
    @Published public var volume: Float = 0.3
    @Published public var isPlaying: Bool = false
    @Published public var isStudyModeEnabled: Bool = false
    
    private var avPlayer: AVAudioPlayer?
    private var fadeTimer: Timer?
    private var isTTSActive: Bool = false
    
    private let presetKey = "vachanam.soundscape.preset"
    private let volumeKey = "vachanam.soundscape.volume"
    private let studyModeKey = "vachanam.soundscape.studyMode"
    
    public override init() {
        super.init()
        loadPersistedSettings()
    }
    
    private func loadPersistedSettings() {
        let defaults = UserDefaults.standard
        if let rawPreset = defaults.string(forKey: presetKey),
           let preset = SoundscapePreset(rawValue: rawPreset) {
            self.currentPreset = preset
        }
        
        if defaults.object(forKey: volumeKey) != nil {
            self.volume = defaults.float(forKey: volumeKey)
        } else {
            self.volume = 0.3
        }
        
        self.isStudyModeEnabled = defaults.bool(forKey: studyModeKey)
    }
    
    // MARK: - Preset Selection
    
    public func selectPreset(_ preset: SoundscapePreset?) {
        guard preset != currentPreset else { return }
        
        let wasPlaying = isPlaying
        stopAmbient(fadeOut: false)
        
        self.currentPreset = preset
        
        if let preset = preset {
            UserDefaults.standard.set(preset.rawValue, forKey: presetKey)
        } else {
            UserDefaults.standard.removeObject(forKey: presetKey)
        }
        
        if preset != nil && (wasPlaying || isStudyModeEnabled || isTTSActive) {
            startAmbient(fadeIn: true)
        }
    }
    
    public func setVolume(_ newVolume: Float) {
        let clamped = max(0.0, min(1.0, newVolume))
        self.volume = clamped
        UserDefaults.standard.set(clamped, forKey: volumeKey)
        
        if isPlaying {
            avPlayer?.volume = clamped
        }
    }
    
    public func setStudyMode(_ enabled: Bool) {
        self.isStudyModeEnabled = enabled
        UserDefaults.standard.set(enabled, forKey: studyModeKey)
        
        if enabled && currentPreset != nil && !isPlaying {
            startAmbient(fadeIn: true)
        } else if !enabled && !isTTSActive && isPlaying {
            stopAmbient(fadeOut: true)
        }
    }
    
    // MARK: - Playback Control
    
    public func startAmbient(fadeIn: Bool = true) {
        guard let preset = currentPreset else { return }
        
        AudioSession.shared.configureSession()
        
        guard let url = urlForPreset(preset) else {
            print("Soundscape audio resource not found: \(preset.resourceName).m4a")
            return
        }
        
        do {
            if avPlayer == nil || avPlayer?.url != url {
                avPlayer = try AVAudioPlayer(contentsOf: url)
                avPlayer?.delegate = self
                avPlayer?.numberOfLoops = -1 // Infinite looping
                avPlayer?.prepareToPlay()
            }
            
            fadeTimer?.invalidate()
            fadeTimer = nil
            
            if fadeIn {
                avPlayer?.volume = 0.0
                avPlayer?.play()
                avPlayer?.setVolume(self.volume, fadeDuration: 1.0)
            } else {
                avPlayer?.volume = self.volume
                avPlayer?.play()
            }
            
            DispatchQueue.main.async {
                self.isPlaying = true
            }
        } catch {
            print("Failed to start ambient player: \(error.localizedDescription)")
            DispatchQueue.main.async {
                self.isPlaying = false
            }
        }
    }
    
    public func pauseAmbient() {
        guard isPlaying else { return }
        avPlayer?.pause()
        DispatchQueue.main.async {
            self.isPlaying = false
        }
    }
    
    public func stopAmbient(fadeOut: Bool = true) {
        guard isPlaying, let player = avPlayer else {
            avPlayer = nil
            DispatchQueue.main.async {
                self.isPlaying = false
            }
            return
        }
        
        fadeTimer?.invalidate()
        fadeTimer = nil
        
        if fadeOut {
            player.setVolume(0.0, fadeDuration: 0.5)
            fadeTimer = Timer.scheduledTimer(withTimeInterval: 0.55, repeats: false) { [weak self] _ in
                player.stop()
                self?.avPlayer = nil
                DispatchQueue.main.async {
                    self?.isPlaying = false
                }
            }
        } else {
            player.stop()
            avPlayer = nil
            DispatchQueue.main.async {
                self.isPlaying = false
            }
        }
    }
    
    // MARK: - TTS Lifecycle Hooks
    
    public func handleTTSPlayStarted() {
        isTTSActive = true
        if currentPreset != nil && !isPlaying {
            startAmbient(fadeIn: true)
        }
    }
    
    public func handleTTSPaused() {
        isTTSActive = false
        if !isStudyModeEnabled && isPlaying {
            pauseAmbient()
        }
    }
    
    public func handleTTSResumed() {
        isTTSActive = true
        if currentPreset != nil && !isPlaying {
            startAmbient(fadeIn: true)
        }
    }
    
    public func handleTTSStopped() {
        isTTSActive = false
        if !isStudyModeEnabled && isPlaying {
            stopAmbient(fadeOut: true)
        }
    }
    
    // MARK: - Helper
    
    private func urlForPreset(_ preset: SoundscapePreset) -> URL? {
        let name = preset.resourceName
        let ext = "m4a"
        
        // 1. Check in Resources/Soundscapes
        if let url = Bundle.main.url(forResource: name, withExtension: ext, subdirectory: "Soundscapes") {
            return url
        }
        // 2. Check in main bundle root
        if let url = Bundle.main.url(forResource: name, withExtension: ext) {
            return url
        }
        // 3. Fallback for test bundles or direct path
        for bundle in [Bundle(for: AmbientSoundscapePlayer.self), Bundle.allBundles.first(where: { $0.bundleIdentifier?.contains("Vachanam") == true })].compactMap({ $0 }) {
            if let url = bundle.url(forResource: name, withExtension: ext, subdirectory: "Soundscapes") {
                return url
            }
            if let url = bundle.url(forResource: name, withExtension: ext) {
                return url
            }
        }
        return nil
    }
}

File: VachanamAndroid/app/src/main/java/com/vachanam/reader/audio/AmbientSoundscapePlayer.kt
package com.vachanam.reader.audio

import android.content.Context
import android.media.MediaPlayer
import com.vachanam.reader.R
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

enum class SoundscapePreset(
    val id: String,
    val title: String,
    val subtitle: String,
    val rawResId: Int
) {
    BROWN_NOISE(
        id = "brown_noise",
        title = "Brown Noise",
        subtitle = "Deep, soothing low-frequency rumble",
        rawResId = R.raw.brown_noise
    ),
    PINK_NOISE(
        id = "pink_noise",
        title = "Pink Noise",
        subtitle = "Balanced waterfall-like calming noise",
        rawResId = R.raw.pink_noise
    ),
    BINAURAL_FOCUS(
        id = "binaural_focus_40hz",
        title = "Binaural Focus",
        subtitle = "40 Hz gamma rhythm for deep focus",
        rawResId = R.raw.binaural_focus_40hz
    ),
    SOFT_RAIN(
        id = "soft_rain",
        title = "Soft Rain",
        subtitle = "Gentle rainfall with droplet ambience",
        rawResId = R.raw.soft_rain
    ),
    LIBRARY_AMBIENCE(
        id = "library_ambience",
        title = "Library & Café",
        subtitle = "Warm, quiet acoustic atmosphere",
        rawResId = R.raw.library_ambience
    );

    companion object {
        fun fromId(id: String): SoundscapePreset? {
            return entries.firstOrNull { it.id.equals(id, ignoreCase = true) }
        }
    }
}

class AmbientSoundscapePlayer(private val context: Context) {
    private val prefs = context.getSharedPreferences("vachanam_soundscape", Context.MODE_PRIVATE)

    private val _currentPreset = MutableStateFlow<SoundscapePreset?>(null)
    val currentPreset: StateFlow<SoundscapePreset?> = _currentPreset.asStateFlow()

    private val _volume = MutableStateFlow(0.3f)
    val volume: StateFlow<Float> = _volume.asStateFlow()

    private val _isPlaying = MutableStateFlow(false)
    val isPlaying: StateFlow<Boolean> = _isPlaying.asStateFlow()

    private val _isStudyModeEnabled = MutableStateFlow(false)
    val isStudyModeEnabled: StateFlow<Boolean> = _isStudyModeEnabled.asStateFlow()

    private var mediaPlayer: MediaPlayer? = null
    private var fadeJob: Job? = null
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())
    private var isTTSActive: Boolean = false

    init {
        loadPersisted()
    }

    private fun loadPersisted() {
        val savedId = prefs.getString("preset", null)
        if (savedId != null) {
            _currentPreset.value = SoundscapePreset.fromId(savedId)
        }
        _volume.value = prefs.getFloat("volume", 0.3f)
        _isStudyModeEnabled.value = prefs.getBoolean("studyMode", false)
    }

    fun selectPreset(preset: SoundscapePreset?) {
        if (_currentPreset.value == preset) return

        val wasPlaying = _isPlaying.value
        stopAmbient(fadeOut = false)

        _currentPreset.value = preset
        if (preset != null) {
            prefs.edit().putString("preset", preset.id).apply()
        } else {
            prefs.edit().remove("preset").apply()
        }

        if (preset != null && (wasPlaying || _isStudyModeEnabled.value || isTTSActive)) {
            startAmbient(fadeIn = true)
        }
    }

    fun setVolume(newVolume: Float) {
        val clamped = newVolume.coerceIn(0.0f, 1.0f)
        _volume.value = clamped
        prefs.edit().putFloat("volume", clamped).apply()
        if (_isPlaying.value) {
            mediaPlayer?.setVolume(clamped, clamped)
        }
    }

    fun setStudyMode(enabled: Boolean) {
        _isStudyModeEnabled.value = enabled
        prefs.edit().putBoolean("studyMode", enabled).apply()

        if (enabled && _currentPreset.value != null && !_isPlaying.value) {
            startAmbient(fadeIn = true)
        } else if (!enabled && !isTTSActive && _isPlaying.value) {
            stopAmbient(fadeOut = true)
        }
    }

    fun startAmbient(fadeIn: Boolean = true) {
        val preset = _currentPreset.value ?: return

        try {
            if (mediaPlayer == null) {
                mediaPlayer = MediaPlayer.create(context, preset.rawResId)?.apply {
                    isLooping = true
                }
            }

            fadeJob?.cancel()

            if (fadeIn) {
                mediaPlayer?.setVolume(0f, 0f)
                mediaPlayer?.start()
                _isPlaying.value = true

                fadeJob = scope.launch {
                    val targetVol = _volume.value
                    val steps = 15
                    for (i in 1..steps) {
                        val v = targetVol * (i.toFloat() / steps)
                        mediaPlayer?.setVolume(v, v)
                        delay(60)
                    }
                }
            } else {
                val v = _volume.value
                mediaPlayer?.setVolume(v, v)
                mediaPlayer?.start()
                _isPlaying.value = true
            }
        } catch (_: Exception) {
            _isPlaying.value = false
        }
    }

    fun pauseAmbient() {
        if (!_isPlaying.value) return
        mediaPlayer?.pause()
        _isPlaying.value = false
    }

    fun stopAmbient(fadeOut: Boolean = true) {
        val player = mediaPlayer ?: run {
            _isPlaying.value = false
            return
        }

        fadeJob?.cancel()

        if (fadeOut) {
            fadeJob = scope.launch {
                val currentVol = _volume.value
                val steps = 10
                for (i in (steps - 1) downTo 0) {
                    val v = currentVol * (i.toFloat() / steps)
                    player.setVolume(v, v)
                    delay(50)
                }
                player.stop()
                player.release()
                mediaPlayer = null
                _isPlaying.value = false
            }
        } else {
            player.stop()
            player.release()
            mediaPlayer = null
            _isPlaying.value = false
        }
    }

    // TTS Lifecycle Hooks
    fun handleTTSPlayStarted() {
        isTTSActive = true
        if (_currentPreset.value != null && !_isPlaying.value) {
            startAmbient(fadeIn = true)
        }
    }

    fun handleTTSPaused() {
        isTTSActive = false
        if (!_isStudyModeEnabled.value && _isPlaying.value) {
            pauseAmbient()
        }
    }

    fun handleTTSStopped() {
        isTTSActive = false
        if (!_isStudyModeEnabled.value && _isPlaying.value) {
            stopAmbient(fadeOut = true)
        }
    }

    companion object {
        @Volatile
        private var instance: AmbientSoundscapePlayer? = null

        fun getInstance(context: Context): AmbientSoundscapePlayer {
            return instance ?: synchronized(this) {
                instance ?: AmbientSoundscapePlayer(context.applicationContext).also { instance = it }
            }
        }
    }
}
